text/0000-global-panic-handler.md

```diff
@@ -0,0 +1,183 @@
+* Handlers would be invoked on nested panics, which would result in a stack
+  overflow if a handler itself panicked.
+* The callbacks were specified to take the panic message, file name and line
+  number directly. This would prevent us from adding more functionality in
+  the future, such as access to backtrace information. In addition, the
+  presence of file names and line numbers for all panics causes some amount of
+  binary bloat and we may want to add some avenue to allow for the omission of
+  those values in the future.
+
+# Detailed design
+
+A new module, `std::panic`, will be created with a panic handling API:
+
+```rust
+/// Unregisters the current panic handler, returning it.
+///
+/// If no custom handler is registered, the default handler will be returned.
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread. Note that this will be a nested
+/// panic and therefore abort the process.
+pub fn take_handler() -> Box<Fn(&PanicInfo) + 'static + Sync + Send> { ... }
+
+/// Registers a custom panic handler, replacing any that was previously
+/// registered.
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread. Note that this will be a nested
+/// panic and therefore abort the process.
+pub fn set_handler<F>(handler: F) where F: Fn(&PanicInfo) + 'static + Sync + Send { ... }
+
+/// A struct providing information about a panic.
+pub struct PanicInfo { ... }
+
+impl PanicInfo {
+    /// Returns the payload associated with the panic.
+    ///
+    /// This will commonly, but not always, be a `&'static str` or `String`.
+    pub fn payload(&self) -> &Any + Send { ... }
+
+    /// Returns information about the location from which the panic originated,
+    /// if available.
+    pub fn location(&self) -> Option<Location> { ... }
+}
+
+/// A struct containing information about the location of a panic.
+pub struct Location<'a> { ... }
+
+impl<'a> Location<'a> {
+    /// Returns the name of the source file from which the panic originated.
+    pub fn file(&self) -> &str { ... }
+
+    /// Returns the line number from which the panic originated.
+    pub fn line(&self) -> u32 { ... }
+}
+```
+
+When a panic occurs, but before unwinding begins, the runtime will call the
+registered panic handler. After the handler returns, the runtime will then
+unwind the thread. If a thread panics while panicking (a "double panic"), the
+panic handler will *not* be invoked and the process will abort. Note that the
+thread is considered to be panicking while the panic handler is running, so a
+panic originating from the panic handler will result in a double panic.
+
+The `take_handler` method exists to allow for handlers to "chain" by closing
+over the previous handler and calling into it:
+
+```rust
+let old_handler = panic::take_handler();
+panic::set_handler(move |info| {
+    println!("uh oh!");
+    old_handler(info);
+});
+```
+
+This is obviously a racy operation, but as a single global resource, the global
```

Comment:
@vadimcn 

Hm that's a good point! It looks like the two primary things there (if I'm reading it right) are:

* Handlers are stacked, not overwritten. This means that everything registered is called in sequence.
* Handlers all have a token, and the token can be used to unregister the handler.

I think that scheme doesn't have the race problems I'm thinking about here, and it also provides the nice ability to "shut down" something and have handlers in theory scoped for smaller components. Additionally, you don't have to "by default try to be robust" because you don't have to worry about callling someone else's handler.

@sfackler, those semantics sound relatively appealing to me, what do you think?

---

@ranma42 

The major downside of a thread-local handler is that you can't set a handler for yet-to-be-spawned threads. I believe that any sort of global handler scheme can be used to implement a thread-local handler scheme (e.g. with more restrictive interfaces).